rfcs/20180918-functions-not-sessions-20.md

```diff
@@ -0,0 +1,1008 @@
+
+with tf.Session() as sess:
+  saver.restore(sess, "/tmp/checkpoint/")
+  sess.run(train_op)</pre>
+
+
+   </td>
+   <td>
+
+
+
+<pre class="prettyprint">W = tf.Variable(
+  tf.glorot_uniform_initializer()(
+    (10, 10)))
+
+@tf.defun
+def train():
+  W.assign_add(1.)
+
+train()
+ckpt = tf.train.Checkpoint(W=W)
+ckpt.save("/tmp/checkpoint")
+ckpt.restore("/tmp/checkpoint")</pre>
+
+
+   </td>
+  </tr>
+  <tr>
+   <td colspan="2" >Exporting/Importing <code>GraphDefs</code>
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+
+
+<pre class="prettyprint">W = tf.get_variable("weights", shape=[10, 10])
+x = tf.placeholder(
+  tf.float32, shape=(None, 10)))
+y = tf.matmul(x, W)
+
+graph = tf.get_default_graph()
+graph_def =  graph.as_graph_def()
+with open("/tmp/graph.pb", "w") as f:
+  f.write(
+      graph_def.SerializeToString())
+
+tf.reset_default_graph()
+
+graph_def = tf.GraphDef()
+with open("/tmp/graph.pbtxt") as f:
+  graph_def.ParseFromString(f.read())
+
+tf.import_graph_def(graph_def)</pre>
+
+
+   </td>
+   <td>
+
+
+
+<pre class="prettyprint">W = tf.Variable(
+  tf.glorot_uniform_initializer()(
+    (10, 10)))
+
+@tf.defun
+def f(x):
+  return tf.matmul(x, W)
+
+# Retrieve the object corresponding to
+# a particular input signature:
+graph = f.graph_function(
+  (tf.float32, (None, 10)).graph
+graph_def = graph.as_graph_def()
+
+with open("/tmp/graph.pb", "w") as f:
+  f.write(graph_def.SerializeToString())</pre>
+
+ \
+
+   </td>
+  </tr>
+  <tr>
+   <td colspan="2" >Exporting/Importing SavedModels
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+
+
+<pre class="prettyprint">
+def save_model():
+  W = tf.get_variable("weights",
+                      shape=[10, 10])
+  x = tf.placeholder(
+    tf.float32, shape=(None, 10))
+  y = tf.matmul(x, W)
+
+  with tf.Session() as sess:
+    sess.run(
+    tf.global_variables_initializer())
+    tf.saved_model.simple_save(
+      sess,
+      "/tmp/model",
+      inputs={"x": x},
+      outputs={"y": y})
+
+def load_model():
+  sess = tf.Session()
+  with sess.as_default():
+    inputs, outputs =  tf.saved_model.simple_load(sess, "/tmp/model")
+  return inputs, outputs, sess</pre>
+
+
+   </td>
+   <td>To be worked on but something along the lines of:
+
+
+
+<pre class="prettyprint">
+class Model(tf.train.Checkpointable):
+  def __init__(self):
+    self.W = tf.Variable(...)
+
+  @tf.method
+  def f(self, x):
+    return tf.matmul(x, self.W)
+
+m = Model()
+
+tf.saved_model.export("/tmp/model", m)
+
+m =
+  tf.saved_model.import("/tmp/model")
+</pre>
+
+
+   </td>
+  </tr>
+</table>
+
+
+
+### Derived/Related Graphs
+
+One reservation expressed by TensorFlow graph/session enthusiasts today is that the ability to write generic analysis/inspection tooling on graphs, precluding the need to understand or modify the Python code that constructed the graph, is important to them. To put it differently, some find it easier to navigate the `GraphDef` program than navigating the Python program. \
```

Comment:
I'm heavily relying on the ability to execute custom subgraphs by using many SessionRunHooks, from what I see here the proposed solution seems unable to give me the flexibility TF 1.0 has.

Consider the following situation:
There are some intermediate tensors/summaries in the graph, which I want to evaluate (i.e. get the results as a numpy array) every once a while but at __different frequency__ (or at other different pre-conditions). Some may need to be evaluated per-step, some every 10k steps. And maybe none will need to be evaluated at all after I finished debugging. These combinations essentially create lot of subgraphs (or the new style "Functions") to evaluate.

With TF1.0 I can easily write SessionRunHooks that evaluate the tensors/summaries I want, at the frequency or other conditions I specify. I can easily modify the tensors/frequencies to evaluate, by editing the list of SessionRunHooks I use. And I can easily disable them all by removing the hooks. By doing that I do not have to touch the code where the tensors/summaries are originally defined (which are likely to be located in multiple files).

What's the alternative here if SessionRunHooks are abandoned?

Reply:
Thanks for the response. Actually I wasn't thinking about using names (strings). String manipulation can easily get wrong so I'm using the tensor objects. A more concrete example:
```python
class Model():
    def my_graph(self, x):
        self.maybe_evaluate = [f(x), g(x), h(x)]
        return output(x)

model = Model()
y = model.my_graph(x)
hooks = [ EvalHook(model.self.maybe_evaluate[0], frequency=10),
        EvalHook(model.self.maybe_evaluate[1], frequency=20), ... ]
while True:
  hooked_session.run(y)
```
Back to your comments: my Python graph construction code is already structured to return these tensors explicitly, because being explicit is less error-prone. 

To use the `NewGraphFunction` for the above flexibility, I assume I would need to make this wrapper (only implementing the outputs part, not the inputs part):
```python
def call_with_hook(f, fetches, hooks):  # f is a defun
    graph = f.graph_function((tf.float32, none)).graph
    fnew = tf.NewGraphFunction(
      graph,
      inputs=[],
      outputs=fetches + hooks.before_run())
    ret = fnew()
    hooks.after_run(ret[len(fetches):])
    return ret[:len(fetches)]
while True:
  call_with_hook(f, y, hooks)
```
and will use this `call_with_hook` wrapper in every iteration.

I can accept this solution, with the hope that `NewGraphFunction` can accept tensor objects 
 in addition to names (presumably that's easy), and it's fast enough (maybe I was just terrified by the name "new" which sounds slow).

Reply:
Something along those lines.
You may still want to cache the updated graph function in each hook (parallel to how any graph construction code is typically in `SessionRunHook.after_create_session` and not `SessionRunHook.before_run`). So more like:

```python
class Hooked(object):
  def __init__(self, f, fetches, hooks):
    graph = f.graph_function((tf.float32, None)).graph
    self.fnew = tf.NewGraphFunction(graph, inputs=[], outputs=hooks.before_run())
    self.hooks = hooks
    self.fetches = fetches

  def __call__(self):
    ret = self.fnew()[len(fetches):]
    self.hooks.after_run(ret)
    return ret

hooked = Hooked(f, y, hooks)
while True:
  hooked()
```

(This is just a rough off-the-top-of-my-head draft demonstrating that such hooking is possible and hopefully intuitive. The finer details may still need some working out)

Reply:
But the point of my use case is that `before_run()` may return different fetches in each iteration and will therefore need to be called at each iteration -- this is also how the hooks work in 1.0. This makes it necessary to create many "GraphFunction" in 2.0, and I'm wondering whether it will have any negative performance issues compared to 1.0.

Reply:
(Resolving as I take @ppwwyyxx 's thumbs-up emoji to mean that this concern has been addressed :)

Reply:
Of course! We do intend to keep and improve various good things in TensorFlow, including the various ecosystem components. Just that this RFC is aimed at the lower level APIs so doesn't get into the details of all the higher level ecosystem. Thanks.

Reply:
That doesn't exist yet, still a TODO but a bit lower down in the priority list at the moment.
(FYI @alextp )

Reply:
tf.graph_function ended up being called tf.compat.v1.wrap_function and the object it returns has a prune method, which does what NewGraphFunction does AFAIU

Reply:
@alextp The name of `compat.v1` means this feature is not considered part of TF 2.0, right?
Are we still going to have this feature in TF 2.0? I would not like to see code full of `tf.compat.v1` because 2.0 lacks features in 1.0.

Reply:
tf.compat.v1 symbols will be usable throughout the lifetime of tf 2.x but will not be kept on 3.x and beyond (and we have no idea when 3.x will come)